components/interactive-markdown.js

```diff
@@ -23,41 +29,101 @@ function ClickToRenderCodePreview(props) {
+
+  const codeBlockHandler = createCodeHandler(StaticCodeBlock)
+  const htmlHandler = highlightHtml(codeBlockHandler)
+  const javaScriptHandler = highlightJavaScript(codeBlockHandler)
+  const bashHandler = highlightBash(codeBlockHandler)
+  const codePreviewHandler = createCodeHandler(CodePreview)
+  const clickToRenderCodePreviewHandler = createCodeHandler(
+    ClickToRenderCodePreview,
+  )
+
   const pragmaHandlers = {
     interactive(options, value) {
-      const component = options.clickToRender ?
-        ClickToRenderCodePreview :
-        CodePreview
-      return {component, code: value, ...options}
+      const interativeCodeBlock = options.clickToRender ?
+        clickToRenderCodePreviewHandler :
+        codePreviewHandler
+      return interativeCodeBlock(options, value)
     },
     callout(options, value) {
       return {component: Callout, children: value, ...options}
     },
+    javascript(options, value) {
+      return javaScriptHandler(options, value)
+    },
+    bash(options, value) {
+      return bashHandler(options, value)
+    },
+    html(options, value) {
+      return htmlHandler(options, value)
+    },
+  }
+
+  return function matcher(specification) {
+    const supportedPragma = Object.keys(pragmaHandlers).find(specification)
+    return pragmaHandlers[supportedPragma] ?
+      pragmaHandlers[supportedPragma] :
+      codePreviewHandler
+  }
+}
+
+function matchPragmaByLanguageName(lang) {
+  return function specification(pragma) {
+    return lang === pragma
+  }
+}
+
+function matchPragmaByLanguageShortName(lang) {
+  const names = {
+    javascript: 'js',
   }
 
+  return function specification(pragma) {
+    return lang === names[pragma]
+  }
+}
+
+function createCodeHandler(component) {
+  return function codeHandler(options, value) {
+    return {component, code: value, ...options}
+  }
+}
+
+function interactiveMarkdown(markdownString) {
+  const componentBlocks = []
+
   function plugin() {
     return ast => {
       visit(ast, 'code', codeNode => {
-        Object.keys(pragmaHandlers).some(pragma => {
-          if (!codeNode.lang || codeNode.lang.indexOf(pragma) !== 0) {
-            return false
-          }
-          const space = 1
-          const options = getOptions(codeNode.lang.slice(pragma.length + space))
-          componentBlocks.push(pragmaHandlers[pragma](options, codeNode.value))
-          // if it's a special code block then we need to
-          // change the node from a code block to a paragraph
-          // so when we do our replacing stuff it doesn't mess up
-          // the HTML
-          Object.assign(codeNode, {
-            type: 'paragraph',
-            children: [{type: 'text', value: 'COMPONENT_BLOCK'}],
-            lang: null,
-            value: null,
-          })
-          return true
+        const language = getLanguage(codeNode.lang)
+        const getMatchingPragma = createSupportedPragmaMatcher()
+        const specification = or(
```

Comment:
I think that I'd rather imperative code here. Even though I'm a fan of FP, I think it'd be more clear imperatively. Either that, or give the term `specification` a better name. What is that?